### ISSUE TITLE
After 1.38.1 Update: Internal error detected. Please copy the above traceback and report

### ISSUE THREAD

**Issue by karrtikr** [microsoft/ptvsd#1780]
@TurboCooler commented on [Mon Sep 16 2019](https://github.com/microsoft/vscode-python/issues/7428)

I can no longer debug.  I get the same error even if I try to debug a simple print statement.  It is like stuck on some cache of files that it cannot forget
```
Traceback (most recent call last):
  File "c:\Users\username\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\ptvsd_launcher.py", line 21, in <module>
    import ptvsd
  File "c:\Users\username\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\__init__.py", line 13, in <module>
    from ._vendored import force_pydevd
  File "c:\Users\username\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\_vendored\force_pydevd.py", line 18, in <module>
    pydevd_constants = import_module('_pydevd_bundle.pydevd_constants')
  File "C:\Python27\lib\importlib\__init__.py", line 37, in import_module
    __import__(name)
  File "c:\Users\username\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\_vendored\pydevd\_pydevd_bundle\pydevd_constants.py", line 205, in <module>
    from _pydev_imps._pydev_saved_modules import thread
  File "c:\Users\username\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\_vendored\pydevd\_pydev_imps\_pydev_saved_modules.py", line 15, in <module>
    import xmlrpclib
  File "C:\Python27\lib\xmlrpclib.py", line 145, in <module>
    import httplib
  File "C:\Python27\lib\httplib.py", line 80, in <module>
    import mimetools
  File "C:\Python27\lib\mimetools.py", line 6, in <module>
    import tempfile
  File "C:\Python27\lib\tempfile.py", line 35, in <module>
    from random import Random as _Random
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\random.py", line 30, in <module>
    from Crypto import Random
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\__init__.py", line 29, in <module>
    from Crypto.Random import _UserFriendlyRNG
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\_UserFriendlyRNG.py", line 38, in <module>
    from Crypto.Random.Fortuna import FortunaAccumulator
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\Fortuna\FortunaAccumulator.py", line 39, in <module>
    import FortunaGenerator
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\Fortuna\FortunaGenerator.py", line 35, in <module>
    from Crypto.Util import Counter
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Util\Counter.py", line 59, in <module>
    from Crypto.Util import _counter
ImportError: cannot import name _counter

Internal error detected. Please copy the above traceback and report at
https://github.com/Microsoft/vscode-python/issues/new

Press Enter to close. . .
```
if it matters, is barfing on counter.so file

---

@karrtikr commented on [Tue Sep 17 2019](https://github.com/microsoft/vscode-python/issues/7428#issuecomment-532325116)

Please search for `Developer: Log database storage contents` in the command palette, and then delete the cache folders. Please see if you can reproduce the issue after that and let us know.

---

@TurboCooler commented on [Tue Sep 17 2019](https://github.com/microsoft/vscode-python/issues/7428#issuecomment-532329205)

> Please search for Developer: Log database storage contents in the command palette, and then delete the cache folders. Please see if you can reproduce the issue after that and let us know.

I found something that says "Developer: Log Storage Database Contents" but choosing that opened what looks like the developer console when you are trying to debug a web page.  When I click on elements it highlights the main VSC screens so it looks like that command debugs the Electron Code and does not give me any folders for a cache.

---

@TurboCooler commented on [Tue Sep 17 2019](https://github.com/microsoft/vscode-python/issues/7428#issuecomment-532331785)

Okay, digging a little more I found the following:
C:\Users\{my-user-name}\AppData\Roaming\Code\User\globalStorage
state.vscdb

And

C:\Users\{my-user-name}\AppData\Roaming\Code\User\workspaceStorage
This folder has some 20 folders named using a GUID

Do I delete some, all of those items I just listed above?

---

@karrtikr commented on [Tue Sep 17 2019](https://github.com/microsoft/vscode-python/issues/7428#issuecomment-532338255)

Yes, actually that's what I meant. Try deleting all contents in `globalStorage` and `workspaceStorage`. Those are the cache folders.

---

@TurboCooler commented on [Tue Sep 17 2019](https://github.com/microsoft/vscode-python/issues/7428#issuecomment-532350395)

I received the same error that I originally posted after I deleted all the cache files you asked me to delete

**Comment by int19h** [comment]
@TurboCooler It sounds like you have "C:\src\merchant_changes_py27\merchant_management\Crypto\Random" set as your current directory ("cwd" in launch.json), or are starting your app as a program from that directory, and so the contents of your package ends up on `sys.path`. Thus, "random.py" inside your package gets exposed as a top-level `random` module, which clashes with, and overrides, the stdlib `random` module that the debugger tries to import.

Can you show your launch.json?

**Comment by TurboCooler** [comment]
If I run from the root path which is c:\src\merchant_changes_py27\merchant_management  I still get the same error.

{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "env": {
                "APEX_FUNCTION_NAME":"merchant_management",
                "AWS_ENVIRONMENT":"qa",
                "CORS_ACCESS_CONTROL_ALLOW_ORIGIN":"{private-url}",
                "CTSURL":"{private-url}",
                "CTS_QUEUE_DELAY_MAP":"{ \"0\": \"300\", \"300\": \"301\", \"301\": \"302\", \"302\": \"303\", \"303\":  \"304\",  \"304\":  \"305\", \"305\":  \"2000\"}",
                "KMSKey":"{private-url}",
                "LAMBDA_FUNCTION_NAME":"{private-url}",
                "OMNI_API_VERSION":"v2",
                "OMNI_Public_Bucket":"{private-url}",
                "{private-data}.GB":"{private-data}",
                "{private-data}.US":"{private-data}",
                "{private-data}":"US, GB",
                "{private-data}.GB":"{private-data}",
                "{private-data}.US":"{private-data}",
                "{private-data}URL":"{private-url}",
                "TrackingURL":"{private-url}",
                "PYTHONPATH" : "C:\\src\\merchant_changes_py27\\merchant_management;C:\\src\\merchant_changes_py27\\merchant_management\\common;C:\\src\\merchant_changes_py27\\merchant_management\\Crypto;C:\\src\\merchant_changes_py27\\merchant_management\\Crypto\\Random;C:\\src\\merchant_changes_py27\\merchant_management\\Crypto\\Random\\Util",
            }
        }
    ]
}

**Comment by TurboCooler** [comment]
Here is the run output from the latest run

cd 'c:\src\merchant_changes_py27\merchant_management'; ${env:APEX_FUNCTION_NAME}='merchant_management'; ${env:AWS_ENVIRONMENT}='qa'; ${env:CORS_ACCESS_CONTROL_ALLOW_ORIGIN}='{private}'; ${env:CTSURL}='{private}'; ${env:CTS_QUEUE_DELAY_MAP}='{ "0": "300", "300": "301", "301": "302", "302": "303", "303":  "304",  "304":  "305", "305":  "2000"}'; ${env:KMSKey}='{private}'; ${env:LAMBDA_FUNCTION_NAME}='{private}'; ${env:OMNI_API_VERSION}='v2'; ${env:OMNI_Public_Bucket}='{private}'; ${env:{private}.GB}='{private}'; ${env:{private}.US}='{private}'; ${env:{private}}='US, GB'; ${env:{private}.GB}='{private}'; ${env:ReboundLogin.US}='Borderfree'; ${env:{private}}='{private}'; ${env:TrackingURL}='{private}'; ${env:PYTHONPATH}='C:\src\merchant_changes_py27\merchant_management;C:\src\merchant_changes_py27\merchant_management\common;C:\src\merchant_changes_py27\merchant_management\Crypto;C:\src\merchant_changes_py27\merchant_management\Crypto\Random;C:\src\merchant_changes_py27\merchant_management\Crypto\Random\Util'; ${env:PYTHONIOENCODING}='UTF-8'; ${env:PYTHONUNBUFFERED}='1'; & 'C:\Python27\python.exe' 'c:\Users\en002ma.PBI\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\ptvsd_launcher.py' '--default' '--client' '--host' 'localhost' '--port' '2788' 'c:\src\merchant_changes_py27\merchant_management\test_merch_call.py'
Traceback (most recent call last):
  File "c:\Users\en002ma.PBI\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\ptvsd_launcher.py", line 21, in <module>
    import ptvsd
  File "c:\Users\en002ma.PBI\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\__init__.py", line 13, in <module>
    from ._vendored import force_pydevd
  File "c:\Users\en002ma.PBI\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\_vendored\force_pydevd.py", line 18, in <module>
    pydevd_constants = import_module('_pydevd_bundle.pydevd_constants')
  File "C:\Python27\lib\importlib\__init__.py", line 37, in import_module
    __import__(name)
  File "c:\Users\en002ma.PBI\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\_vendored\pydevd\_pydevd_bundle\pydevd_constants.py", line 205, in <module>
    from _pydev_imps._pydev_saved_modules import thread
  File "c:\Users\en002ma.PBI\.vscode\extensions\ms-python.python-2019.9.34911\pythonFiles\lib\python\ptvsd\_vendored\pydevd\_pydev_imps\_pydev_saved_modules.py", line 15, in <module>
    import xmlrpclib
  File "C:\Python27\lib\xmlrpclib.py", line 145, in <module>
    import httplib
  File "C:\Python27\lib\httplib.py", line 80, in <module>
    import mimetools
  File "C:\Python27\lib\mimetools.py", line 6, in <module>
    import tempfile
  File "C:\Python27\lib\tempfile.py", line 35, in <module>
    from random import Random as _Random
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\random.py", line 30, in <module>
    from Crypto import Random
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\__init__.py", line 29, in <module>
    from Crypto.Random import _UserFriendlyRNG
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\_UserFriendlyRNG.py", line 38, in <module>
    from Crypto.Random.Fortuna import FortunaAccumulator
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\Fortuna\FortunaAccumulator.py", line 39, in <module>
    import FortunaGenerator
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Random\Fortuna\FortunaGenerator.py", line 35, in <module>
    from Crypto.Util import Counter
  File "C:\src\merchant_changes_py27\merchant_management\Crypto\Util\Counter.py", line 59, in <module>
    from Crypto.Util import _counter
ImportError: cannot import name _counter

Internal error detected. Please copy the above traceback and report at
https://github.com/Microsoft/vscode-python/issues/new

Press Enter to close. . .

**Comment by karthiknadig** [comment]
You have path to `C:\src\merchant_changes_py27\merchant_management\Crypto\Random` added in PYTHONPATH. Python picks up `random.py` from that folder, when it should be picking from the standard lib.

I don't think you need to add any of that to PYTHONPATH. You can set the `cwd` to be ${workspaceFolder}. That should be enough.

**Comment by TurboCooler** [comment]
> You have path to C:\src\merchant_changes_py27\merchant_management\Crypto\Random added in PYTHONPATH. Python picks up random.py from that folder, when it should be picking from the standard lib.
> I don't think you need to add any of that to PYTHONPATH. You can set the cwd to be ${workspaceFolder}. That should be enough.

That Random is our own Random.  Not part of the standard library

**Comment by TurboCooler** [comment]
If I do not use those path then VSC does not pick up all the libraries that it should I get an error that module is not found

**Comment by int19h** [comment]
From Python's perspective, there  is no distinction between standard library and your code. If you have a folder that contains random.py on PYTHONPATH, and somebody does `import random`, it will get loaded. Since `random` is a standard module, this effectively breaks any Python package that you use directly or indirectly that depends on `random`.

Can you show some of those imports that are failing for you without those PYTHONPATH entries? Since you have `__init__.py` inside Crypto, it is a package. In general, you shouldn't need to add paths *inside* a package to your PYTHONPATH, and doing so is likely to break things. A module from it should be loaded by something like `import Crypto.Random` or `from Crypto import Random`, and this should work fine with only "C:\src\merchant_changes_py27\merchant_management" on PYTHONPATH (and even then, you need to add it explicitly only if it's not already the current directory).

**Comment by TurboCooler** [comment]
Okay, so I removed the PYTYHONPATH.  The debugger no longer crashes.  It stops at the same error but that error is on me.  it is a compiled C program with a python bridge named counter.so  

However, I did previously add the PYTHONPATH specifically to address the fact that I was not able to resolve the imports.  Now the imports are being resolved.  

Definitely this specific issue here is resolved but I also definitely was struggling with imports not being found previously.  Not sure if clearing the cache as I was previously asked to do resolved that problem.  However, I was also NOT crashing before applying 1.38.1

I am not saying I am not at fault here with what I did with PYTHONPATH, but at the same time I was not applying the PYTHONPATH because I felt like it.

Thanks everyone who has responded to help me address the issue.

Exception has occurred: ImportError
cannot import name _counter
  File "C:\omnipath\merchant_changes_py27\merchant_management\Crypto\Util\Counter.py", line 59, in <module>
    from Crypto.Util import _counter
  File "C:\omnipath\merchant_changes_py27\merchant_management\Crypto\Random\Fortuna\FortunaGenerator.py", line 35, in <module>
    from Crypto.Util import Counter
  File "C:\omnipath\merchant_changes_py27\merchant_management\Crypto\Random\Fortuna\FortunaAccumulator.py", line 39, in <module>
    import FortunaGenerator
  File "C:\omnipath\merchant_changes_py27\merchant_management\Crypto\Random\_UserFriendlyRNG.py", line 38, in <module>
    from Crypto.Random.Fortuna import FortunaAccumulator
  File "C:\omnipath\merchant_changes_py27\merchant_management\Crypto\Random\__init__.py", line 29, in <module>
    from Crypto.Random import _UserFriendlyRNG
  File "C:\omnipath\merchant_changes_py27\merchant_management\common\kms.py", line 3, in <module>
    from Crypto import Random
  File "C:\omnipath\merchant_changes_py27\merchant_management\common\dynamo.py", line 10, in <module>
    import kms
  File "C:\omnipath\merchant_changes_py27\merchant_management\main.py", line 3, in <module>
    from common import dynamo
  File "C:\omnipath\merchant_changes_py27\merchant_management\test_merch_call.py", line 1, in <module>
    import main

**Comment by karthiknadig** [comment]
@TurboCooler I am happy that we were able to unblock you partially at least :) . Please reach out to us if there are any more issues. Closing this for now.
